### ISSUE TITLE
Physics informed neural network (PINN)

### ISSUE THREAD

**Issue by geokefe** [liufeng2317/TorchInversion#7]
# -----------------------------------
    #     Define DIP model
    # -----------------------------------
    model_shape = [nz,nx]
    DIP_model = DIP_CNN(model_shape,in_channels=[16,32,16],vmin=vp_true.min()/1000,vmax=vp_true.max()/1000,device=device)
    DIP_model.to(device)

    # -----------------------------------
    #     Pretrain DIP model
    # -----------------------------------
    pretrain        = True
    load_pretrained = False
    if pretrain:
        if load_pretrained:
            # load the model parameters
            DIP_model.load_state_dict(torch.load(os.path.join(project_path,f"inversion-{layer_num}layer-16-32-16/DIP_model_pretrained.pt")))
        else:
            lr          = 0.005
            iteration   = 10000
            step_size   = 1000
            gamma       = 0.5
            optimizer = torch.optim.Adam(DIP_model.parameters(),lr = lr)
            scheduler = torch.optim.lr_scheduler.StepLR(optimizer,step_size=step_size,gamma=gamma)
            vp_init = numpy2tensor(vp_init,dtype=dtype).to(device)
            pbar = tqdm(range(iteration+1))
            for i in pbar:  
                vp_nn = DIP_model()
                loss = torch.sqrt(torch.sum((vp_nn - vp_init)**2))
                optimizer.zero_grad()
                loss.backward()
                optimizer.step()
                scheduler.step()
                pbar.set_description(f'Pretrain Iter:{i}, Misfit:{loss.cpu().detach().numpy()}')
            torch.save(DIP_model.state_dict(),os.path.join(project_path,f"inversion-{layer_num}layer-16-32-16/DIP_model_pretrained.pt"))

    I'm trying to modify the CNN module (typical of the code above) in ADFWI to work for PINN, I need guide on how to effectively do this and also modify the loss function to account for model penalty of interest.

**Comment by liufeng2317** [comment]
Actually, the code above is not a Physics-Informed Neural Network (PINN). Rather, it leverages a deep representation akin to the Deep Image Prior (DIP). Specifically, we utilize a neural network where the inputs are random numbers, and the outputs are velocity models. To prevent the random initialization from causing the Full Waveform Inversion (FWI) to fall into local minima, we perform an end-to-end mapping, where the network transforms the random inputs into a known initial velocity model. After this, the FWI proceeds to fine-tune the network parameters. For more details, please refer to the following article:

Liu, F., Li, H., Zou, G., & Li, J. (2024). Automatic Differentiation-based Full Waveform Inversion with Flexible Workflows. arXiv preprint arXiv:2412.00486.

**Comment by geokefe** [comment]
I know the code is not a PINN, however, I'm trying to explore the possibility of modifying it to incorporate PINN algorithm. I just want to know the feasibility of this modification and get possible guide.
